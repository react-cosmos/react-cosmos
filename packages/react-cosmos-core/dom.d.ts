export * from './dist/dom/index.js';
