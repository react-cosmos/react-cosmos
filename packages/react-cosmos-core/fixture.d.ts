export * from './dist/fixture/index.js';
