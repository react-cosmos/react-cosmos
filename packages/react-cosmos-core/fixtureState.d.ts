export * from './dist/fixtureState/index.js';
