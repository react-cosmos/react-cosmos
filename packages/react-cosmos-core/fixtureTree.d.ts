export * from './dist/fixtureTree/index.js';
