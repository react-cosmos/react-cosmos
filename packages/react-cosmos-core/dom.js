module.exports = require('./dist/dom/index.js');
