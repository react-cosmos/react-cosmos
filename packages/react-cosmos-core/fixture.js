module.exports = require('./dist/fixture/index.js');
