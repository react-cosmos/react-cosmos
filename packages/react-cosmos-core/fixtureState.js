module.exports = require('./dist/fixtureState/index.js');
