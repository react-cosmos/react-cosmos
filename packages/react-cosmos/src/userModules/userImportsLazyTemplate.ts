import { flatten } from 'lodash-es';
import {
  UserImportsTemplateArgs,
  createImportMap,
  importPath,
} from './shared.js';

export function userImportsLazyTemplate({
  globalImports,
  modulePaths: { fixturePaths, decoratorPaths },
  rendererConfig,
  rootDir,
  relativeToDir,
  typeScript,
}: UserImportsTemplateArgs) {
  function ext(filePath: string) {
    // Add compat with projects where allowImportingTsExtensions isn't enabled
    return typeScript ? filePath.replace(/\.tsx?$/, '') : filePath;
  }

  function ts(annotation: string) {
    return typeScript ? annotation : '';
  }

  const globalImportItems = globalImports.map(
    p => `import '${ext(importPath(p, relativeToDir))}';`
  );

  const fixtures = createImportMap(fixturePaths, rootDir, relativeToDir);
  const fixtureKeys = Object.keys(fixtures);
  const fixtureItems = fixtureKeys.map(
    k => `  '${k}': { getModule: () => import('${ext(fixtures[k])}') }`
  );

  const decorators = createImportMap(decoratorPaths, rootDir, relativeToDir);
  const decoratorKeys = Object.keys(decorators);
  const decoratorItems = decoratorKeys.map(
    k => `  '${k}': { getModule: () => import('${ext(decorators[k])}') }`
  );

  const rendererConfigStr = JSON.stringify(rendererConfig, null, 2);

  let importGroups = [globalImportItems];
  if (typeScript) {
    importGroups.unshift([
      `import { RendererConfig, UserModuleWrappers } from 'react-cosmos-core';`,
    ]);
  }

  return `
// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.
${importsStr(importGroups)}
export const rendererConfig${ts(': RendererConfig')} = ${rendererConfigStr};

const fixtures = {${objBody(fixtureItems)}};

const decorators = {${objBody(decoratorItems)}};

export const moduleWrappers${ts(': UserModuleWrappers')} = {
  lazy: true,
  fixtures,
  decorators
};
`.trimStart();
}

function importsStr(items: string[][]) {
  if (flatten(items).length === 0) {
    return '';
  }

  const str = items
    .filter(rows => rows.length > 0)
    .map(rows => rows.join('\n'))
    .join('\n\n');

  return `\n${str}\n`;
}

function objBody(items: string[]) {
  return items.length > 0 ? `\n${items.join(`,\n`)}\n` : '';
}
