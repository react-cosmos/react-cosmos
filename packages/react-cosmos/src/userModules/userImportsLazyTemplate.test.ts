import { RendererConfig } from 'react-cosmos-core';
import { userImportsLazyTemplate } from './userImportsLazyTemplate.js';

const globalImports = [
  '/Users/ovidiu/cosmos/src/polyfills.ts',
  '/Users/ovidiu/cosmos/src/global.css',
];

const fixturePaths = [
  '/Users/ovidiu/cosmos/src/Counter/Counter.fixture.tsx',
  '/Users/ovidiu/cosmos/src/CounterButton/CounterButton.fixture.tsx',
];

const decoratorPaths = ['/Users/ovidiu/cosmos/src/cosmos.decorator.tsx'];

const modulePaths = { fixturePaths, decoratorPaths };

const rendererConfig: RendererConfig = {
  playgroundUrl: 'http://localhost:5002',
};

it('should generate user imports with absolute paths', () => {
  expect(
    userImportsLazyTemplate({
      rootDir: '/Users/ovidiu/cosmos',
      modulePaths,
      globalImports,
      rendererConfig,
      relativeToDir: null,
      typeScript: false,
    })
  ).toMatchSnapshot();
});

it('should generate user imports with relative paths', () => {
  expect(
    userImportsLazyTemplate({
      rootDir: '/Users/ovidiu/cosmos',
      modulePaths,
      globalImports,
      rendererConfig,
      relativeToDir: '/Users/ovidiu/cosmos/src',
      typeScript: false,
    })
  ).toMatchSnapshot();
});

it('should generate TypeScript user imports with absolute paths', () => {
  expect(
    userImportsLazyTemplate({
      rootDir: '/Users/ovidiu/cosmos',
      modulePaths,
      globalImports,
      rendererConfig,
      relativeToDir: null,
      typeScript: true,
    })
  ).toMatchSnapshot();
});

it('should generate TypeScript user imports with relative paths', () => {
  expect(
    userImportsLazyTemplate({
      rootDir: '/Users/ovidiu/cosmos',
      modulePaths,
      globalImports,
      rendererConfig,
      relativeToDir: '/Users/ovidiu/cosmos/src',
      typeScript: true,
    })
  ).toMatchSnapshot();
});
