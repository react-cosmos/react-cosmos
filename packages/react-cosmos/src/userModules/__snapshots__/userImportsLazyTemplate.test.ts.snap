// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate TypeScript user imports with absolute paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

import { RendererConfig, UserModuleWrappers } from 'react-cosmos-core';

import '/Users/ovidiu/cosmos/src/polyfills';
import '/Users/ovidiu/cosmos/src/global.css';

export const rendererConfig: RendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
  'src/Counter/Counter.fixture.tsx': { getModule: () => import('/Users/ovidiu/cosmos/src/Counter/Counter.fixture') },
  'src/CounterButton/CounterButton.fixture.tsx': { getModule: () => import('/Users/ovidiu/cosmos/src/CounterButton/CounterButton.fixture') }
};

const decorators = {
  'src/cosmos.decorator.tsx': { getModule: () => import('/Users/ovidiu/cosmos/src/cosmos.decorator') }
};

export const moduleWrappers: UserModuleWrappers = {
  lazy: true,
  fixtures,
  decorators
};
"
`;

exports[`should generate TypeScript user imports with relative paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

import { RendererConfig, UserModuleWrappers } from 'react-cosmos-core';

import './polyfills';
import './global.css';

export const rendererConfig: RendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
  'src/Counter/Counter.fixture.tsx': { getModule: () => import('./Counter/Counter.fixture') },
  'src/CounterButton/CounterButton.fixture.tsx': { getModule: () => import('./CounterButton/CounterButton.fixture') }
};

const decorators = {
  'src/cosmos.decorator.tsx': { getModule: () => import('./cosmos.decorator') }
};

export const moduleWrappers: UserModuleWrappers = {
  lazy: true,
  fixtures,
  decorators
};
"
`;

exports[`should generate user imports with absolute paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

import '/Users/ovidiu/cosmos/src/polyfills.ts';
import '/Users/ovidiu/cosmos/src/global.css';

export const rendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
  'src/Counter/Counter.fixture.tsx': { getModule: () => import('/Users/ovidiu/cosmos/src/Counter/Counter.fixture.tsx') },
  'src/CounterButton/CounterButton.fixture.tsx': { getModule: () => import('/Users/ovidiu/cosmos/src/CounterButton/CounterButton.fixture.tsx') }
};

const decorators = {
  'src/cosmos.decorator.tsx': { getModule: () => import('/Users/ovidiu/cosmos/src/cosmos.decorator.tsx') }
};

export const moduleWrappers = {
  lazy: true,
  fixtures,
  decorators
};
"
`;

exports[`should generate user imports with relative paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

import './polyfills.ts';
import './global.css';

export const rendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
  'src/Counter/Counter.fixture.tsx': { getModule: () => import('./Counter/Counter.fixture.tsx') },
  'src/CounterButton/CounterButton.fixture.tsx': { getModule: () => import('./CounterButton/CounterButton.fixture.tsx') }
};

const decorators = {
  'src/cosmos.decorator.tsx': { getModule: () => import('./cosmos.decorator.tsx') }
};

export const moduleWrappers = {
  lazy: true,
  fixtures,
  decorators
};
"
`;
